File: real_estate_fe/app/types/rbac.ts
/** Role trong hệ thống RBAC (GET /admin/roles) */
export interface RoleResponse {
  id: number
  /** Mã role: CUSTOMER / BROKER / ADMIN */
  code: string
  name: string
  description: string
  is_active: boolean
  /** Danh sách permission code thuộc role này */
  permissions: string[]
}

/** Permission trong hệ thống RBAC (GET /admin/permissions) */
export interface PermissionResponse {
  id: number
  /** Mã permission, ví dụ: user.read */
  code: string
  name: string
  /** Nhóm chức năng mà permission thuộc về */
  module: string
  description: string
}

/** Người dùng ở góc nhìn quản trị (GET /admin/users) */
export interface AdminUserResponse {
  id: number
  name: string
  email: string
  phone: string
  is_active: boolean
  /** Danh sách role code đang gán cho user */
  roles: string[]
  created_at: string
}

/**
 * Tham số truy vấn danh sách user cho admin.
 * Dùng type alias (không dùng interface) để gán được vào QueryParams của BaseService.
 */
export type AdminUserQuery = {
  search?: string
  page?: number
  size?: number
}

File: real_estate_be/internal/dto/rbac.go
package dto

// ── RBAC: role / permission ──────────────────────────────

// RoleResponse — 1 role kèm danh sách permission được gán
type RoleResponse struct {
	ID          uint64   `json:"id"`
	Code        string   `json:"code"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	IsActive    bool     `json:"is_active"`
	Permissions []string `json:"permissions"`
}

type PermissionResponse struct {
	ID          uint64 `json:"id"`
	Code        string `json:"code"`
	Name        string `json:"name"`
	Module      string `json:"module"`
	Description string `json:"description"`
}

// AdminUserResponse — 1 user trong màn quản trị, kèm các role đang giữ
type AdminUserResponse struct {
	ID        uint64   `json:"id"`
	Name      string   `json:"name"`
	Email     string   `json:"email"`
	Phone     string   `json:"phone"`
	IsActive  bool     `json:"is_active"`
	Roles     []string `json:"roles"`
	CreatedAt string   `json:"created_at"`
}

// SetUserRolesRequest — gán lại toàn bộ role của user (một user có thể giữ nhiều role)
type SetUserRolesRequest struct {
	Roles []string `json:"roles"`
}
